### ISSUE TITLE
$matches sporadic error message PossiblyUndefinedVariable

### ISSUE THREAD

**Issue by timglabisch** [vimeo/psalm#4748]
sometimes we've issues with error messages like:

```
OK ERROR: PossiblyUndefinedVariable - xxxx.php:49:21 - Possibly undefined variable $matches defined in try block (see https://psalm.dev/018)
OK         $x = (int)$matches[1];
```

the code looks similar like this:
https://psalm.dev/r/46e568f7f8

i could not reproduce this issue in psalm.dev - complaining about "PossiblyUndefinedVariable" in this case seems to be sporadic. Some of my co-workers sometimes have the issue.

**Comment by orklah** [comment]
Could you check if it's not sporadic anymore when running psalm with --no-diff and/or --no-cache please?

Could you also post a more complete example in psalm.dev? The issue warns that your preg_match is in a try{} block

**Comment by timglabisch** [comment]
@orklah we are running it with no cache `php ./vendor/bin/psalm --no-cache --threads=8`. even dropping the cache folder (`rm -rf ~/.cache/psalm`) doesn't help.

**Comment by pilif** [comment]
I see the same issue related to the `$output` argument to `exec`. It's also different depending on thread count. I had the most luck reproducing it with --threads=4, but it doesn't always happen and also, it started happening after completely unrelated changes to the code-base.

**Comment by DvdGiessen** [comment]
I'm having the same issue with a `PossiblyUndefinedVariable` being raised for a `$matches` output of `preg_match`.

Running with `--no-cache` and explicitly deleted cache directories.

I have multiple instances where `preg_match` is used in this way in multiple places in my codebase. It doesn't detect for every instance of this in my codebase consistently. In my case, it appears to detects the same instances when running multiple times, as long as the codebase is unchanged. If the codebase is changed (also when changing files other files than the one the error is in), sometimes not all instances of this `preg_match` usage detect errors anymore, or instances that previously weren't detected start being detected.

I tried making a minimal testcase with just this code but due to the inconsistency mentioned I was unable to make it reproduce consistently using a small test file containing just this code. :(

Noteworthy observation: it also differers between operating systems. When on a real Linux machine, a Linux VM and WSLv1 it doesn't detects it, while on a Windows machine it does, or the other way round. It always gets detected on just one: either on Windows or on a flavour of Linux, not on both. When changing the codebase as mentioned above, it may change which OS detects certain instances. Not sure if there's just a timing difference between these OSes, or this might hint towards another cause.

Example of error:
```php
40:            } elseif (preg_match('/^b\'([^\']+)\'$/', $data, $matches)) {
41:                $data = stripcslashes($matches[1]);
42:            } elseif (preg_match('/^b"([^"]+)"$/', $data, $matches)) {
43:                $data = stripcslashes($matches[1]);
43:            }
```
```
ERROR: PossiblyUndefinedVariable - _snip_.php:41:39 - Possibly undefined variable $matches defined in try block
ERROR: PossiblyUndefinedVariable - _snip_.php:42:59 - Possibly undefined variable $matches defined in try block
ERROR: PossiblyUndefinedVariable - _snip_.php:43:39 - Possibly undefined variable $matches defined in try block
```

For now, I've resorted to just ignoring `PossiblyUndefinedVariable` across my whole code base. (Since it is a generic `IssueHandlerType` I can't filter on `referencedVariable="$matches"` or something like that.)

**Comment by etremblay** [comment]
We are having the same issue with a large code base.

psalm sometimes report PossiblyUndefinedVariable on a file when scanning the whole project, but not on sa single file.

I run psalm with about 10 threads

**Comment by pilif** [comment]
this is still happening to us and it's completely random when it starts happening. Once it does start, it's happening independently of the `--threads` -setting.

I think this is dependent on evaluation order.

In our case, it's most often occurring around the `&$output` parameter of `exec`

**Comment by orklah** [comment]
I'm pretty sure I found the issue here and fixed it in #6553 

The fix I made explain every observation we made here:
- It seems to trigger on params with reference
- It shows after a try catch
- changing number of threads can change the order of analysis and hide the issue in some case
- however it has no impact on @DvdGiessen who has the two calls in the same file

If someone who can reproduce the issue reliably could try to run Psalm analysis with patch from #6553 it would be most helpful. Otherwise, we'll consider this fixed once PR is merged.

**Comment by DvdGiessen** [comment]
Quickly tried it on my local machine; with the patch applied my issue no longer reproduces. As noted before I was seeing some non-determinism, so I'll keep an eye on it when the next Psalm release with this change is out.
